File: R/the-name-game.R
#!/usr/bin/Rscript --slave

# Come on everybody
# I say now let's play a game
# I betcha I can make a rhyme
# Out of anybody's name

# R is probably not the best language for text manipulation, but it is
# workable.

combine <- function(letter, name) {
  # Combine a letter and a name according to the rules of the name game

  first <- tolower(substring(name, 1, 1));
  stem <- substring(name, 2);

  if (grepl("[aeiou]", first)) {
    # Vowel rule: eg 'b' + "Emily" -> 'bemily"
    return(paste(letter, first, stem, sep=""));
  }
  else if (grepl("^[cht]", stem)) {
    # Two initial consonants: drop both, eg 'b' + "Shirley" -> "birley"
    return(paste(letter, substring(stem, 2), sep=""));
  }
  else if (grepl("[bfm]", letter) && letter == first) {
    # Letter matches first of name rule: eg 'm' + "Maggie" -> "aggie"
    # That's the only rule that is contrary
    return(stem);
  }
  else {
    # Normal rule: eg 'f' + "Rupert" -> "fupert"
    return(paste(letter, stem, sep=""));
  }
}

rhymeName <- function(name) {
  # Rhyme name according to the rules of the name game

  if (nchar(name) < 2 || ! grepl("^[a-zA-Z]+$", name)) {
    return(paste(name, "isn't a name, so I can't rhyme\n"));
  }

  return(paste(name, ", ", name, ", ",
               "bo-", combine('b', name), "\n",
               "Banana-fana fo-", combine('f', name), "\n",
               "Fee-fi mo-", combine('m', name), "\n",
               name, "!\n",
               sep=""));
}

# Execution starts here

argv <- commandArgs(TRUE)
if (length(argv) != 1) {
  cat("usage: the-name-game NAME\n")
  q(status = 1)
}

cat(rhymeName(argv[1]));

File: js/the-name-game.js
#!/usr/bin/env nodejs

// Come on everybody
// I say now let's play a game
// I betcha I can make a rhyme
// Out of anybody's name

// Combine LETTER and NAME, eg b and Ann to bann
function combine(letter, name) {
    var first = name.charAt(0).toLowerCase();
    var stem = name.substring(1);

    if (first.match(/[aeiou]/)) {
        // Vowel rule: eg 'b' + "Emily" -> 'bemily"
        return letter + first + stem;
    } else if (stem.charAt(0).match(/[cht]/)) {
        // Two initial consonants: drop both, eg 'b' + "Shirley" -> "birley"
        return letter + stem.substring(1);
    } else if (first.match(/[bmf]/) && letter == first) {
        // Letter matches first of name rule: eg 'm' + "Maggie" -> "aggie"
        // That's the only rule that is contrary
        return stem;
    } else {
        // Normal rule: eg 'f' + "Rupert" -> "fupert"
        return letter + stem;
    }
}

// Check if name can be rhymed and if so constructs the rhyme.
function rhyme_name(name) {
    if (name.length < 2 || ! name.match(/^[a-z]+$/i)) {
        return name + " isn't a name, so I can't rhyme";
    }

    return name + ", " + name + ", bo-" + combine("b", name) + "\n" +
        "Banana-fana fo-" + combine("f", name) + "\n" +
        "Fee-fi mo-" + combine("m", name) + "\n" +
        name + "!";
}

if (process.argv.length < 3) {
    console.log("usage: the-name-game NAME");
    process.exit(1);
}

console.log(rhyme_name(process.argv[2]));
